### PR TITLE
【Hackathon 6th No.8】NO.8 为 Paddle 新增 FeatureAlphaDropout API

### PR DESCRIPTION
**Pull request by megemini**
### PR Category

User Experience

### PR Types

Docs

### Description

[NO.8 为 Paddle 新增 FeatureAlphaDropout API](https://github.com/PaddlePaddle/community/blob/master/hackathon/hackathon_6th/%E3%80%90Hackathon%206th%E3%80%91%E5%BC%80%E6%BA%90%E8%B4%A1%E7%8C%AE%E4%B8%AA%E4%BA%BA%E6%8C%91%E6%88%98%E8%B5%9B%E6%A1%86%E6%9E%B6%E5%BC%80%E5%8F%91%E4%BB%BB%E5%8A%A1%E5%90%88%E9%9B%86.md#no8-%E4%B8%BA-paddle-%E6%96%B0%E5%A2%9E-featurealphadropout-api)

相关接口的 RFC

参考实现代码 https://github.com/PaddlePaddle/Paddle/pull/64881

### CODE REVIEW

File: rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+  if (!alpha_dropout) {
+    return multiply<inplace>(input, noise);
+  } else {
+    return multiply<inplace>(input, noise).add_(b);
+  }
+}
+```
+
+可以看到，由于 `feature alpha dropout` 为 `alpha dropout` 的一种特殊情况，因此， PyTorch 将两者统一实现在以上的同一个函数中，两者的唯一区别为，`feature alpha dropout` 的 `noise` 是通过 `make_feature_noise` 生成的，而不是 `alpha dropout` 中的一个与输入相同形状的空张量：
```
Comment (zxcd) [inline rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md:75]: 这里通过 make_feature_noise 应该是 feature_dropout，而非 feature alpha dropout，这块需要重新调研一下。另外`feature alpha dropout` 为 `alpha dropout` 的一种特殊情况这点并不确切，_dropout_impl 包含了4种 dropout，alpha dropout 是 dropout 的特殊情况，你需要再看一下 feature alpha dropout 的具体分支

File: rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+def feature_alpha_dropout(x, p=0.5, training=True, name=None):
+    return _feature_alpha_dropout_impl(
+        x, feature_dropout=True, p=p, training=training, name=name
+    )
+```
+
+其唯一区别为 `feature_dropout` 的参数不同。
+
+## 命名与参数设计
```
Comment (zxcd) [inline rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md:195]: 这块写一下

File: rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+                raise ValueError(
+                    'Feature alpha dropout needs at least 2D input.'
+                )
+            input_shape = list(x.shape[:2]) + [1] * len(x.shape[2:])
+
+```
+
+# 六、测试和验收的考量
+
+- **编程范式场景**
```
Comment (zxcd) [inline rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md:327]: 增加测试文件存放路径

File: rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+  - 常规覆盖动态图 (和静态图) 的测试场景。
+
+- **硬件场景**
+  - 常规需覆盖 CPU、GPU 两种测试场景。
+
+- **输入参数**
+  - 常规覆盖默认参数，常用参数，错误参数。
+  - 常规数据类型 float16, float32 or float64
```
Comment (zxcd) [inline rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md:335]: 需要支持bfloat16

File: rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md (hunk 1 above)
Comment (megemini) [reply rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md:75]: 我之前描述的太笼统了，我是假定：只关心我们要实现 `feature_alpha_dropout` ，这个前提来描述的 ～

不好意思，这里再展开说一下，帮忙看看对不对 ～

torch 的 `feature_alpha_dropout` 是这里的 `_dropout_impl` 的一种特定实现：

`aten/src/ATen/functorch/PyTorchOperatorHacks.cpp`

``` c++
ALIAS_SPECIALIZATION(_dropout,               false, false)
ALIAS_SPECIALIZATION(_feature_dropout,       true,  false)
ALIAS_SPECIALIZATION(_alpha_dropout,         false, true )
ALIAS_SPECIALIZATION(_feature_alpha_dropout, true,  true )

#define ALIAS_SPECIALIZATION(ALIAS_NAME, IS_FEATURE, IS_ALPHA)                      \
template <bool inplace, typename... Args>                                           \
Ctype<inplace> ALIAS_NAME(Args&&... args) {                                         \
  return _dropout_impl<IS_FEATURE, IS_ALPHA, inplace>(std::forward<Args>(args)...); \
}
```

也就是说，`feature_alpha_dropout` 可以看成是这里的 `dropout + alpha_dropout + feature_dropout` 或者说是 `alpha_dropout(含有 dropout 的含义) + feature_dropout`。其次，`feature_alpha_dropout` 相对于 `alpha_dropout`  相当于 `dropout` 的 `范围` 更具体 （`alpha_dropout` 是对每个点，`feature_alpha_dropout` 是针对某个通道），因此，我这里说 `feature_alpha_dropout` 是 `alpha_dropout` 的一种特殊情况 ～

我这里说 

> `feature alpha dropout` 的 `noise` 是通过 `make_feature_noise` 生成的

的原因是，我这里只关心 `feature alpha dropout` 而不是 `feature dropout`，因此，看作 `feature alpha dropout 的 noise 是通过 make_feature_noise 生成的` ～ 这里是我之前描述的不严谨 ～

其实，我更关心的是，我文中的实现方案是否有问题？

文中的方案针对：

``` c++
ALIAS_SPECIALIZATION(_dropout,               false, false)
ALIAS_SPECIALIZATION(_feature_dropout,       true,  false)
ALIAS_SPECIALIZATION(_alpha_dropout,         false, true )
ALIAS_SPECIALIZATION(_feature_alpha_dropout, true,  true )
```

只涉及后面两种，直接复用 paddle 的 `alpha_dropout` ，只是 `input_shape` 有区别 ～

我觉得应该可以，不知道是否有问题？

谢谢！

File: rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md (hunk 3 above)
Comment (megemini) [reply rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md:327]: paddle 目前把 `alpha_dropout` 等 dropout 都放在 `test/legacy_test/test_dropout_op.py` 中 ～ 

我这里也这样处理，可否？

File: rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md (hunk 4 above)
Comment (megemini) [reply rfcs/APIs/20240604_api_design_for_feature_alpha_dropout.md:335]: 简单测试了一下，应该可以：

``` python
In [14]: feature_alpha_dropout(i_paddle, 0.2, training=True)
Out[14]: 
Tensor(shape=[4, 3, 3], dtype=bfloat16, place=Place(gpu:0), stop_gradient=True,
       [[[ 0.80859375,  1.07812500,  0.83203125],
         [ 0.73437500,  0.33398438,  0.96484375],
         [ 0.72656250,  1.16406250,  0.54687500]],

        [[ 0.71093750,  0.50781250,  1.15625000],
         [-1.23437500, -1.23437500, -1.23437500],
         [-1.23437500, -1.23437500, -1.23437500]],

        [[ 0.34960938,  0.34375000,  0.98046875],
         [ 0.78125000,  0.83203125,  0.98828125],
         [ 0.64843750,  1.11718750,  0.37109375]],

        [[-1.23437500, -1.23437500, -1.23437500],
         [ 0.48437500,  0.31250000,  0.60156250],
         [ 0.81640625,  1.00781250,  1.09375000]]])

```

**Review by zxcd — APPROVED** [review]
LGTM
